File: gateway/internal/session/session.go
package session

import (
	"sync"
	"time"

	"github.com/google/uuid"
)

type Session struct {
	ID        string            `json:"id"`
	Channel   string            `json:"channel"`
	UserID    string            `json:"user_id"`
	History   []Turn            `json:"history"`
	Metadata  map[string]string `json:"metadata,omitempty"`
	CreatedAt time.Time         `json:"created_at"`
	UpdatedAt time.Time         `json:"updated_at"`
}

type Turn struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type Manager struct {
	mu       sync.RWMutex
	sessions map[string]*Session
}

func NewManager() *Manager {
	return &Manager{
		sessions: make(map[string]*Session),
	}
}

func (m *Manager) GetOrCreate(channel, userID string) *Session {
	key := channel + ":" + userID

	m.mu.RLock()
	s, ok := m.sessions[key]
	m.mu.RUnlock()

	if ok {
		return s
	}

	m.mu.Lock()
	defer m.mu.Unlock()

	if s, ok := m.sessions[key]; ok {
		return s
	}

	s = &Session{
		ID:        uuid.New().String(),
		Channel:   channel,
		UserID:    userID,
		History:   make([]Turn, 0),
		Metadata:  make(map[string]string),
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	m.sessions[key] = s
	return s
}

func (m *Manager) Get(sessionID string) (*Session, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	s, ok := m.sessions[sessionID]
	return s, ok
}

func (m *Manager) AddTurn(sessionID, role, content string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	if s, ok := m.sessions[sessionID]; ok {
		s.History = append(s.History, Turn{Role: role, Content: content})
		s.UpdatedAt = time.Now()
	}
}

File: core/src/agent/session.rs
use crate::agent::{SessionId, TurnId};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Session {
    pub id: SessionId,
    pub agent_id: String,
    pub channel: String,
    pub user_id: String,
    pub history: Vec<Turn>,
    pub metadata: HashMap<String, String>,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Turn {
    pub id: TurnId,
    pub role: TurnRole,
    pub content: String,
    pub tool_calls: Vec<ToolCallRecord>,
    pub timestamp: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TurnRole {
    User,
    Assistant,
    System,
    Tool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCallRecord {
    pub tool_name: String,
    pub arguments: serde_json::Value,
    pub result: serde_json::Value,
    pub duration_ms: u64,
}

impl Session {
    pub fn new(id: SessionId, channel: String, user_id: String) -> Self {
        Self {
            id,
            agent_id: String::from("default"),
            channel,
            user_id,
            history: Vec::new(),
            metadata: HashMap::new(),
            created_at: chrono::Utc::now(),
            updated_at: chrono::Utc::now(),
        }
    }

    pub fn add_turn(&mut self, user_msg: &str, assistant_msg: &str) {
        use uuid::Uuid;
        self.history.push(Turn {
            id: Uuid::new_v4().to_string(),
            role: TurnRole::User,
            content: user_msg.to_string(),
            tool_calls: Vec::new(),
            timestamp: chrono::Utc::now(),
        });
        self.history.push(Turn {
            id: Uuid::new_v4().to_string(),
            role: TurnRole::Assistant,
            content: assistant_msg.to_string(),
            tool_calls: Vec::new(),
            timestamp: chrono::Utc::now(),
        });
        self.updated_at = chrono::Utc::now();
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentState {
    pub agent_id: String,
    pub name: String,
    pub persona: String,
    pub config: serde_json::Value,
}
